File: src/main/java/carnage/actionLogger/listeners/PlayerActionListeners.java
package carnage.actionLogger.listeners;

import carnage.actionLogger.models.LogEntry;
import carnage.actionLogger.queue.LogQueue;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.*;
import org.bukkit.event.player.*;

public class PlayerActionListeners implements Listener {
    private final LogQueue queue;

    public PlayerActionListeners(LogQueue queue) { this.queue = queue; }

    @EventHandler public void onPlace(BlockPlaceEvent e) { log(e.getPlayer().getName(), "Block Placed", e.getBlock().getType().name(), e.getBlock().getLocation()); }
    @EventHandler public void onBreak(BlockBreakEvent e) { log(e.getPlayer().getName(), "Block Broken", e.getBlock().getType().name(), e.getBlock().getLocation()); }
    @SuppressWarnings("deprecation")
    @EventHandler public void onChat(AsyncPlayerChatEvent e) { log(e.getPlayer().getName(), "Chat", e.getMessage(), e.getPlayer().getLocation()); }
    @EventHandler public void onCommand(PlayerCommandPreprocessEvent e) { log(e.getPlayer().getName(), "Command", e.getMessage(), e.getPlayer().getLocation()); }

    private void log(String p, String a, String d, org.bukkit.Location l) {
        queue.addLog(new LogEntry(p, a, d, l));
    }
}

File: src/main/kotlin/cat/daisy/core/events/PlayerActionListeners.kt
package cat.daisy.core.events

import cat.daisy.core.service.ActionLogService
import cat.daisy.core.service.ActionLogService.Action
import cat.daisy.core.service.ConfigService
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerCommandPreprocessEvent

/**
 * Listens for player block placement/breaking, chat, and command events.
 * Submits logs to ActionLogService for async batch processing.
 */
class PlayerActionListeners(
    private val logs: ActionLogService,
) : Listener {
    private inline fun shouldLog(toggle: String): Boolean =
        ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "enabled", true) &&
            ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", toggle, true)

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onPlace(event: BlockPlaceEvent) {
        if (!shouldLog("capture.block_place")) return

        logs.submit(
            ActionLogService.LogEntry(
                playerName = event.player.name,
                action = Action.BLOCK_PLACED,
                detail = event.block.type.name,
                location = event.block.location,
            ),
        )
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onBreak(event: BlockBreakEvent) {
        if (!shouldLog("capture.block_break")) return

        logs.submit(
            ActionLogService.LogEntry(
                playerName = event.player.name,
                action = Action.BLOCK_BROKEN,
                detail = event.block.type.name,
                location = event.block.location,
            ),
        )
    }

    @Suppress("DEPRECATION") // AsyncPlayerChatEvent deprecation is for Paper API reasons
    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onChat(event: AsyncPlayerChatEvent) {
        if (!shouldLog("capture.chat")) return

        logs.submit(
            ActionLogService.LogEntry(
                playerName = event.player.name,
                action = Action.CHAT,
                detail = event.message,
                location = event.player.location,
            ),
        )
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onCommand(event: PlayerCommandPreprocessEvent) {
        if (!shouldLog("capture.command")) return

        logs.submit(
            ActionLogService.LogEntry(
                playerName = event.player.name,
                action = Action.COMMAND,
                detail = event.message,
                location = event.player.location,
            ),
        )
    }
}
